### PR TITLE
implement structured streaming datahub source

### PR DESCRIPTION
**Pull request by wenxuanguan**
## What changes were proposed in this pull request?

support datahub source in  structured streaming #226 

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

### CODE REVIEW

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  def fetchNextBatch(): Unit = {
+    endOffset = datahubClient.getCursor(project, topic, shardId, CursorType.LATEST).getSequence
+    val topicResult = datahubClient.getTopic(project, topic)
+    val schema = topicResult.getRecordSchema
+    val limit = endOffset - nextOffset
+    val cursor = datahubClient.getCursor(project, topic, shardId, nextOffset).getCursor
+    val recordResult = datahubClient.getRecords(project, topic, shardId, cursor, limit.toInt, schema)
+    val num = recordResult.getRecordCount
+    recordResult.getRecords.foreach(record => {
+      record.getSystemTime
```
Comment (uncleGen) [inline emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala]: ??

**Comment by uncleGen** [comment]
This branch is out-of-date with the base branch

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala [hunk 2]
```diff
@@ -170,8 +165,19 @@ class DatahubContinuousDataReader(
     rowWriter.write(1, UTF8String.fromString(topic))
     rowWriter.write(2, UTF8String.fromString(shardId))
     rowWriter.write(3, DateTimeUtils.fromJavaTimestamp(
-      new java.sql.Timestamp(currentRecord.get(DatahubSchema.SYSTEM_TIME).asInstanceOf[Long])))
-    rowWriter.write(4, currentRecord.toJSONString.getBytes)
+      new java.sql.Timestamp(currentRecord.getSystemTime)))
+
+    fields.getOrElse({
+      fields = Some(datahubClient.getTopic(project, topic).getRecordSchema.getFields.toList)
+      fields.get
```
Comment (uncleGen) [inline emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala]: do a schema request to server for each record?

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala (hunk 1 above)
Comment (wenxuanguan) [reply emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala]: unused, and deleted

File: emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala (hunk 2 above)
Comment (wenxuanguan) [reply emr-datahub/src/main/scala/org/apache/spark/sql/aliyun/datahub/DatahubContinuousReader.scala]: no, the fields in schema will request from server and be stored when handling first record, and will reuse for the rest of records.

**Pull request merged**